vm/actor/src/builtin/miner/expiration_queue.rs

```diff
@@ -0,0 +1,785 @@
+                rescheduled_power += &group.power;
+            }
+
+            self.must_update_or_delete(group.epoch, expiration_set)?;
+        }
+
+        if !sectors_total.is_empty() {
+            // Add sectors to new expiration as early-terminating and faulty.
+            let early_sectors: BitField = sectors_total.iter().map(|&i| i as usize).collect();
+            self.add(
+                new_expiration,
+                &BitField::new(),
+                &early_sectors,
+                &PowerPair::zero(),
+                &rescheduled_power,
+                &TokenAmount::zero(),
+            )?;
+        }
+
+        Ok(&rescheduled_power + &expiring_power)
+    }
+
+    /// Re-schedules *all* sectors to expire at an early expiration epoch, if they wouldn't expire before then anyway.
+    pub fn reschedule_all_as_faults(
+        &mut self,
+        fault_expiration: ChainEpoch,
+    ) -> Result<(), Box<dyn StdError>> {
+        let mut rescheduled_epochs = Vec::<u64>::new();
+        let mut rescheduled_sectors = BitField::new();
+        let mut rescheduled_power = PowerPair::zero();
+
+        let mut mutated_expiration_sets = Vec::<(ChainEpoch, ExpirationSet)>::new();
+
+        self.amt.for_each(|e, expiration_set| {
+            let epoch = e as ChainEpoch;
+
+            if epoch <= self.quant.quantize_up(fault_expiration) {
+                let mut expiration_set = expiration_set.clone();
+
+                // Regardless of whether the sectors were expiring on-time or early, all the power is now faulty.
+                // Pledge is still on-time.
+                expiration_set.faulty_power += &expiration_set.active_power;
+                expiration_set.active_power = PowerPair::zero();
+                mutated_expiration_sets.push((epoch, expiration_set));
+            } else {
+                rescheduled_epochs.push(epoch as u64);
+                rescheduled_sectors |= &expiration_set.on_time_sectors;
+                rescheduled_sectors |= &expiration_set.early_sectors;
+                rescheduled_power += &expiration_set.active_power;
+                rescheduled_power += &expiration_set.faulty_power;
+            }
+
+            Ok(())
+        })?;
+
+        for (epoch, expiration_set) in mutated_expiration_sets {
+            self.must_update(epoch, expiration_set)?;
+        }
+
+        // If we didn't reschedule anything, we're done.
+        if rescheduled_epochs.is_empty() {
+            return Ok(());
+        }
+
+        // Add rescheduled sectors to new expiration as early-terminating and faulty.
+        self.add(
+            fault_expiration,
+            &BitField::new(),
+            &rescheduled_sectors,
+            &PowerPair::zero(),
+            &rescheduled_power,
+            &TokenAmount::zero(),
+        )?;
+
+        // Trim the rescheduled epochs from the queue.
+        for epoch in rescheduled_epochs {
+            self.amt.delete(epoch)?;
+        }
```

Comment:
noting that if you do change to batch delete, to also switch this one (doesn't have a TODO so I'm commenting)